arch/x86/core/intel64/cpu.c

```diff
@@ -168,15 +81,13 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	}
 }
 
-/* Per-CPU initialization, C domain. On the first CPU, z_x86_prep_c is the
- * next step. For other CPUs it is probably smp_init_top().
- */
 FUNC_NORETURN void z_x86_cpu_init(struct x86_cpuboot *cpuboot)
 {
+	uint8_t cpu_num;
+
 	x86_sse_init(NULL);
 
-	/* The internal cpu_number is the index to x86_cpuboot[] */
-	unsigned char cpu_num = (unsigned char)(cpuboot - x86_cpuboot);
+	cpu_num = cpuboot->cpu_id;
```

Comment:
since you move the variable declaration up to the code black start, then remove the next empty line (as if is testing the value of that assigned variable, which rule does not apply if the variable is declared and assigned at the same time)

Anyway, are you sure there is no misalignment of cpu_id and apic_id (as used in arch_start_cpu) ?

So far the cpu_num was not related to any hardware identifier, for the good reason that sometime there are CPUs with N cores but only K cores are actually valid and activated. But the inactivated cores can still be listed and owning a legit apic_id/cpu_id, and it's only through ACPI that these can be ignored (thus the logic behind acpi_local_apic_get(): it ignores the disabled ones).
Check this, make sure the cpu_id you are getting matches the same core which acip_id will be retrieved from ACPI, when there are inactive cores.